### PR TITLE
move rvalue checking to MIR

### PR DESCRIPTION
**Pull request by arielb1**

### CODE REVIEW

File: src/librustc_mir/transform/type_check.rs [hunk 1]
```diff
@@ -641,9 +649,39 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
...
+                // - maybe we should make that a warning.
+                return
+            }
+            LocalKind::Var | LocalKind::Temp => {}
+        }
+
+        let span = local_decl.source_info.span;
+        let ty = local_decl.ty;
+        if !ty.is_sized(self.tcx().global_tcx(), self.infcx.param_env(), span) {
+            if let None = self.reported_errors.replace((ty, span)) {
```
Comment (eddyb) [inline src/librustc_mir/transform/type_check.rs]: `replace` confuses me - `insert` wouldn't work?

File: src/librustc_mir/transform/type_check.rs [hunk 2]
```diff
@@ -641,9 +649,39 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
...
+        }
+    }
+
+    fn typeck_mir(&mut self, mir: &Mir<'gcx>) {
         self.last_span = mir.span;
         debug!("run_on_mir: {:?}", mir.span);
+
+        for (local, local_decl) in mir.local_decls.iter_enumerated() {
+            self.check_local(mir, local, local_decl);
+        }
```
Comment (eddyb) [inline src/librustc_mir/transform/type_check.rs]: Wait, can't there be assignments between, say, two dereferences? Is there any guarantee there would be a `Local`? If so, a comment would be helpful.

**Review by eddyb — APPROVED** [review]
r=me, with some clarifications.

File: src/librustc_mir/transform/type_check.rs (hunk 2 above)
Comment (arielb1) [inline src/librustc_mir/transform/type_check.rs]: @bors r=eddyb

**Comment by nikomatsakis** [comment]
@arielb1 nice to see things moving to MIR :)

**Pull request merged**